trie/database.go

```diff
@@ -424,39 +426,63 @@ func (db *Database) Node(hash common.Hash) ([]byte, error) {
 	memcacheDirtyMissMeter.Mark(1)
 
 	// Content unavailable in memory, attempt to retrieve from disk
-	enc, err := db.diskdb.Get(hash[:])
-	if err == nil && enc != nil {
+	enc := rawdb.ReadTrieNode(db.diskdb, hash)
+	if len(enc) != 0 {
 		if db.cleans != nil {
 			db.cleans.Set(hash[:], enc)
 			memcacheCleanMissMeter.Mark(1)
 			memcacheCleanWriteMeter.Mark(int64(len(enc)))
 		}
+		return enc, nil
 	}
-	return enc, err
+	return nil, errors.New("not found")
 }
 
 // preimage retrieves a cached trie node pre-image from memory. If it cannot be
 // found cached, the method queries the persistent database for the content.
-func (db *Database) preimage(hash common.Hash) ([]byte, error) {
+func (db *Database) preimage(hash common.Hash) []byte {
 	// Retrieve the node from cache if available
 	db.lock.RLock()
 	preimage := db.preimages[hash]
 	db.lock.RUnlock()
 
 	if preimage != nil {
-		return preimage, nil
+		return preimage
 	}
-	// Content unavailable in memory, attempt to retrieve from disk
-	return db.diskdb.Get(secureKey(hash))
+	return rawdb.ReadPreimage(db.diskdb, hash)
 }
 
-// secureKey returns the database key for the preimage of key (as a newly
-// allocated byte-slice)
-func secureKey(hash common.Hash) []byte {
-	buf := make([]byte, secureKeyLength)
-	copy(buf, secureKeyPrefix)
-	copy(buf[secureKeyPrefixLength:], hash[:])
-	return buf
+// preimage retrieves a cached contract code from memory. If it cannot be
+// found cached, the method queries the persistent database for the content.
+func (db *Database) Code(hash common.Hash) ([]byte, error) {
+	// Retrieve the contract code from the clean cache if available
+	if db.cleans != nil {
+		if enc := db.cleans.Get(nil, hash[:]); enc != nil {
+			memcacheCleanHitMeter.Mark(1)
+			memcacheCleanReadMeter.Mark(int64(len(enc)))
+			return enc, nil
+		}
+	}
+	// Retrieve the node from cache if available
+	db.lock.RLock()
+	code := db.codes[hash]
+	db.lock.RUnlock()
+
+	if code != nil {
+		memcacheCodeHitMeter.Mark(1)
+		memcacheCodeReadMeter.Mark(int64(len(code)))
+		return code, nil
+	}
+	code = rawdb.ReadCode(db.diskdb, hash)
+	if len(code) > 0 {
+		if db.cleans != nil {
+			db.cleans.Set(hash[:], code)
```

Comment:
Interesting question: If we want to separate code and state, should we use the un-prefixed `hash` as storage key for `cleans`? A couple of variants: 

1. Yes, store by hash in `cleans` (what this PR does)
2. No, store prefixed in `cleans`.
3. No don't store code in `cleans` at all

Reply:
I think it's fine to use the `un-prefixed` hash. 
The reason is even there are two same code and trie node, the hash is same and they are mapped into the same entry, it's fine since the content is **correct**.

The main issue for mixing the code and trie node is: it can interrupt the fast sync if we accidentally sync the code with same hash as the non-retrieved-node.

But it's ok here.